develop/tutorials/articles/290-configuration/02-custom-system-settings-ui.markdown

```diff
@@ -0,0 +1,221 @@
+    level), you might want to exclude it from the System Settings UI.
+
+## Providing Custom Configuration Forms
+
+This method relies on an existing Config Admin configuration class, as described
+[here](/develop/tutorials/-/knowledge_base/7-1/making-applications-configurable). 
+Here's an example configuration class, from Liferay's own Currency Converter
+application:
+
+    @ExtendedObjectClassDefinition(category = "localization")
+    @Meta.OCD(
+        id = "com.liferay.currency.converter.web.configuration.CurrencyConverterConfiguration",
+        localization = "content/Language",
+        name = "currency-converter-configuration-name"
+    )
+    public interface CurrencyConverterConfiguration {
+
+        @Meta.AD(deflt = "GBP|CNY|EUR|JPY|USD", name = "symbols", required = false)
+        public String[] symbols();
+
+    }
+
+There's one configuration option, symbols, that takes an array of values. 
+
+All that's necessary is to customize the auto-generated form is one additional
+class, an implementation of the `ConfigurationFormRenderer` interface.
+
+Implement it's three methods:
+
+1.  `getPid`: Return the id of the configuration object as defined in the `id`
+    property of the `@Meta.OCD` annotation of the `*Configuration` class.
+
+2.  `getRequestParameters`: Read the parameters sent by the custom form and put
+    them in a Map whose keys should be the names of the fields of the
+    Configuration interface.
+
+3.  `render`: Render the custom form's fields, using your desired method (for
+    example, JSPs or another template mechanism). The `<form>` tag itself is
+    provided automatically and shouldn't be included in the
+    `ConfigurationFormRenderer`.
+
+Here's a complete `ConfigurationFormRenderer` implementation:
+
+        @Component(immediate = true, service = ConfigurationFormRenderer.class)
+        public class CurrencyConverterConfigurationFormRenderer
+            implements ConfigurationFormRenderer {
+
+            @Override
+            public String getPid() {
+
+                return "com.liferay.currency.converter.web.configuration.CurrencyConverterConfiguration";
+            }
+
+            @Override
+            public void render(HttpServletRequest request, HttpServletResponse response)
+                throws IOException {
+
+                String formHtml = "<input name=\"mysymbols\" />";
+
+                PrintWriter writer = response.getWriter();
+
+                writer.print(formHtml);
+
+            }
+
+            @Override
+            public Map<String, Object> getRequestParameters(
+                HttpServletRequest request) {
+
+                Map<String, Object> params = new HashMap<>();
+
+                String[] mysymbols = ParamUtil.getParameterValues(request, "mysymbols");
+
+                params.put("symbols", mysymbols);
+
+                return params;
+            }
+        }
+
+The above example ...
```

Comment:
@jorgeferrer What does this example do?

Reply:
It generates a custom rendering (HTML) for the form (in render()) and reads the information entered in the form (in getRequestParameters())